homeassistant/components/system_bridge/coordinator.py

```diff
@@ -68,174 +69,139 @@ def __init__(
                 self.title,
                 exception,
             )
-            if self.unsub:
-                self.unsub()
-                self.unsub = None
-            self.last_update_success = False
-            self.async_update_listeners()
-
-    async def _setup_websocket(self) -> None:
-        """Use WebSocket for updates."""
-        try:
-            async with asyncio.timeout(20):
-                await self.websocket_client.connect()
-
-            self.hass.async_create_background_task(
-                self._listen_for_data(),
-                name="System Bridge WebSocket Listener",
-            )
-
-            await self.websocket_client.register_data_listener(
-                RegisterDataListener(modules=MODULES)
-            )
-            self.last_update_success = True
-            self.async_update_listeners()
-        except AuthenticationException as exception:
-            self.logger.error(
-                "Authentication failed at setup for %s: %s", self.title, exception
-            )
-            if self.unsub:
-                self.unsub()
-                self.unsub = None
-            self.last_update_success = False
-            self.async_update_listeners()
-            raise ConfigEntryAuthFailed(
-                translation_domain=DOMAIN,
-                translation_key="authentication_failed",
-                translation_placeholders={
-                    "title": self.title,
-                    "host": self._host,
-                },
-            ) from exception
-        except ConnectionErrorException as exception:
-            self.logger.warning(
-                "Connection error occurred for %s. Will retry: %s",
-                self.title,
-                exception,
-            )
-            self.last_update_success = False
-            self.async_update_listeners()
-        except TimeoutError as exception:
-            self.logger.warning(
-                "Timed out waiting for %s. Will retry: %s",
-                self.title,
-                exception,
-            )
-            self.last_update_success = False
-            self.async_update_listeners()
-
-        async def close_websocket(_) -> None:
-            """Close WebSocket connection."""
-            await self.websocket_client.close()
-
-        # Clean disconnect WebSocket on Home Assistant shutdown
-        self.unsub = self.hass.bus.async_listen_once(
-            EVENT_HOMEASSISTANT_STOP, close_websocket
-        )
+            await self.clean_disconnect()
 
     async def _async_update_data(self) -> SystemBridgeData:
         """Update System Bridge data from WebSocket."""
-        self.logger.debug(
-            "_async_update_data - WebSocket Connected: %s",
-            self.websocket_client.connected,
-        )
-        if not self.websocket_client.connected:
-            await self._setup_websocket()
+        if self.listen_task is None or not self.websocket_client.connected:
```

Comment:
Please be aware that this function can run after init, because people can use the update_entity service for this.